lib/posticle/src/regexes.rs

```diff
@@ -0,0 +1,149 @@
+#![allow(dead_code)]
+
+use regex::{Regex, RegexBuilder};
+
+static CTRL_CHARS: &str = r"\x00-\x1F\x7F";
+static INVALID_CHARS: &str = "\u{FFFE}\u{FEFF}\u{FFFF}\u{202A}-\u{202E}";
+static UNICODE_SPACES: &str = "\u{09}-\u{0D}\u{20}\u{85}\u{A0}\u{1680}\u{180E}\u{2000}-\u{200A}\u{2028}\u{2029}\u{202F}\u{205F}\u{3000}";
+static LATIN_ACCENTS: &str = "\u{00C0}-\u{00D6}\u{00D8}-\u{00F6}\u{00F8}-\u{00FF}\u{0100}-\u{024F}\u{0253}-\u{0254}\
+    \u{0256}-\u{0257}\u{0259}\u{025b}\u{0263}\u{0268}\u{026F}\u{0272}\u{0289}\u{02BB}\u{1E00}-\u{1EFF}";
+static PUNCTUATION: &str = r##"-_!"#$%&'()*+,./:;<=>?@\[\]^`{|}~"##;
+static PUNCTUATION_NO_HYPHEN: &str = r##"_!"#$%&'()*+,./:;<=>?@\[\]^`{|}~"##;
+static PUNCTUATION_NO_HYPHEN_UNDERSCORE: &str = r##"!"#$%&'()*+,./:;<=>?@\[\]^`{|}~"##;
+static VALID_QUERY_STRING: &str =
+    r##"(?:[-a-zA-Z0-9!?*'\(\);:&=+$/%#\[\]_\.,~|@]*[a-zA-Z0-9_&=#/])"##;
+
+lazy_static! {
+    /// Matches characters which can validly start or end a domain segment.
+    static ref VALID_DOMAIN_EXTREMA_CHARS: String = format!(
+        "[^{punctuation}{ctrl}{invalid}{space}]",
+        punctuation = PUNCTUATION,
+        ctrl = CTRL_CHARS,
+        invalid = INVALID_CHARS,
+        space = UNICODE_SPACES
+    );
+
+    /// Matches characters which are valid in the middle of a domain segment.
+    static ref VALID_DOMAIN_MIDDLE_CHARS: String = format!(
+        "[^{punctuation}{ctrl}{invalid}{space}]",
+        punctuation = PUNCTUATION_NO_HYPHEN,
+        ctrl = CTRL_CHARS,
+        invalid = INVALID_CHARS,
+        space = UNICODE_SPACES
+    );
+
+    /// Matches a valid domain segment.
+    static ref VALID_DOMAIN_SEGMENT: String = format!(
+        "(?:{extremum}(?:{middle}*{extremum})?)",
+        extremum=*VALID_DOMAIN_EXTREMA_CHARS,
+        middle=*VALID_DOMAIN_MIDDLE_CHARS,
+    );
+
+    /// Matches things that _look_ like TLDs (using the very naive heuristic of "two or more valid characters").
```

Comment:
from [RFC 3696](https://tools.ietf.org/html/rfc3696#section-2)
>The LDH rule, as updated, provides that the labels (words or strings
   separated by periods) that make up a domain name must consist of only
   the ASCII [ASCII] alphabetic and numeric characters, plus the hyphen.
   No other symbols or punctuation characters are permitted, nor is
   blank space.  If the hyphen is used, it is not permitted to appear at
   either the beginning or end of a label.  There is an additional rule
   that essentially requires that top-level domain names not be all-
   numeric.
Essentially, it only allows for alphanumeric names plus hyphens (cannot be the first or last character). Non-ASCII domains are stored as punycode, and the domain (when punyencoded if need be) can be up to 63 octets long. There is no restriction on minimum length, and while there are no single-character domains in the root as of now, FreeNIC does have `.o`.

Since we are parsing domain names in raw form, we probably should only pay attention to these rules:
-  Length is 1-63 bytes
- Hyphens are allowed except for the start and end of the domain
- Numbers are allowed anywhere, but the name cannot be numeric-only

Reply:
Hmmmm, how is that different from what we're doing currently? I was approximating what `twitter-text` (plus Mastodon's tweaks to it) do as much as possible.

Reply:
Or, rather
1. Yes, we can implement that; not super sure how important that is
2. That's implemented
3. Also implemented

Reply:
Right now it looks like our regex for a syntactically valid TLD is simply "two or more allowed characters", but my suggestion is to restrict that a bit more to avoid false matches, and also allow single-character domains (if this would have an adverse effect on parsing efficiency, feel free to shoot me down)

Reply:
I suppose my point is mainly that TLDs with hyphens are *allowed* and could be added in the future (and there are alternate DNS roots that may add such TLDs in the future)

Reply:
Oh, yeah, alternate DNS roots are a very good point.